### PR TITLE
Smallstep step 0.8.5 (new formula)

### PR DESCRIPTION
**Pull request by dopey**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by claui** [comment]
Please go away, little bot.

**Comment by dopey** [comment]
Hey @claui thanks for removing the stale tag! Wondering if there's anything maintainers might be waiting on from me to support the PR?

File: Formula/step.rb
Comment (zbeekman) [inline Formula/step.rb]: If you're installing both the Step Cli and the `step-ca` tool, would it be easy/straightforward to test `#{bin}/step-ca` as well?

Also, can you please check for the creation of the key pair files from this test, not just the return status?

File: Formula/step.rb
Comment (zbeekman) [inline Formula/step.rb]: Would it make sense to change the homepage to https://smallstep.com since the formula name is simply `step` and it provides two cli utilities from two separate git repositories?

**Review by zbeekman — CHANGES_REQUESTED** [review]
Looks pretty good to me. I think the tests need further improvement (nothing major). I also wonder if the https://smallstep.com website might be a more appropriate home page since it appears that the build process effectively installs from two separate git repositories.

I'd appreciate some more feedback on this from other @Homebrew/core maintainers who are more familiar with go, and as to whether or not `step` and `step-ca` should be split.

**Comment by zbeekman** [comment]
Cool looking tools!

File: Formula/step.rb
Comment (dopey) [reply Formula/step.rb]: Added a couple of more tests with a few checking for the files that should have been created.

The `step-ca` command kicks off a server which the `cli` (or any other client) can then access via an API. Is kicking off a server from the test-suite acceptable? If so, then we can try delivering a dummy configuration file with the homebrew package, kicking up the server, and making a health request (or something more involved).

Comment (javian) [reply Formula/step.rb]: Yes it is acceptable and there are several examples in the Formulae on how to do it. If you grep for `fork` you should be able to find plenty of examples.

File: Formula/step.rb
Comment (zbeekman) [inline Formula/step.rb]: As noted above, you can launch a server for a test (just make sure tests don't take toooooooooo long). In general we prefer tests to include more than simply `--version`. Please replace the version check with something more appropriate.

File: Formula/step.rb
Comment (zbeekman) [inline Formula/step.rb]: Rather than shelling out, can you please use

```suggestion
     assert_predicate testpath/"foo.priv", :exist?
     assert_predicate testpath/"foo.pub", :exist?
```

File: Formula/step.rb
Comment (zbeekman) [inline Formula/step.rb]: Again, please don't shell out for stuff like this. See above.

File: Formula/step.rb
Comment (zbeekman) [inline Formula/step.rb]: Please add a real test for the functionality of `step-ca`, not a simple version check. Running up a lightweight server test is fine.

File: Formula/step.rb
Comment (zbeekman) [inline Formula/step.rb]: Again, we shouldn't be shelling out & firing up unneeded processes/relying on the behaviour of `grep` and `sed`. Instead can you please use `assert_match` in combination with `shell_output()`? You can look at [avimetaedit.rb L32](https://github.com/Homebrew/homebrew-core/blob/master/Formula/avimetaedit.rb#L32) or [fribidi.rf L29](https://github.com/Homebrew/homebrew-core/blob/master/Formula/fribidi.rb#L29) for examples using `assert_match`.

Comment (zbeekman) [reply Formula/step.rb]: Also, can you please store the result of `shell_output(...)` and then run `assert_match` on that output, so we don't have to run `step certificate inspect baz.crt` twice.

File: Formula/step.rb
Comment (zbeekman) [inline Formula/step.rb]: Same deal here:
1. Don't shell out to `ls`
2. Save the output of `shell_output()` as a variable
3. Use `assert_match` to check the contents of the output for the desired strings.

**Review by zbeekman — CHANGES_REQUESTED** [review]
Almost there, I think!

A few things in the tests need to be cleaned up. Shelling out needs to be reduced/avoided, as discussed in the inline comments. Also, `step-ca` needs a better test other than checking the version. Spinning up a light weight service/server is fine.

File: Formula/step.rb
Comment (MikeMcQuaid) [inline Formula/step.rb]: What does this provide that `https://github.com/smallstep/cli/archive/v0.8.4-rc.4.tar.gz` does not?

Comment (zbeekman) [reply Formula/step.rb]: @MikeMcQuaid It appears to me that it's just a tarball containing https://github.com/smallstep/cli AND https://github.com/smallstep/certificates and then the formula tells go to build and install both.

Comment (zbeekman) [reply Formula/step.rb]: The release ships only the compiled binary, and only for cli.

Comment (MikeMcQuaid) [reply Formula/step.rb]: I think it would be preferable to set the URL to a repo tarball and add the certificates as a `resource` rather than have a special Homebrew tarball.

Comment (zbeekman) [reply Formula/step.rb]: Yeah, I agree, that would make it clearer to maintainers what exactly is going on. The inception double-repo tarball is a bit opaque/mystifying.

File: Formula/step.rb
Comment (dopey) [reply Formula/step.rb]: Thanks for the tip! Added a few tests with a forked server.

File: Formula/step.rb
Comment (dopey) [reply Formula/step.rb]: Added a few tests with a forked server. Should be too long :)

File: Formula/step.rb
Comment (dopey) [reply Formula/step.rb]: Replaced as many of the `shell_output(...)`s as I could and stored the results so I wouldn't have to run `inspect` multiple times.

File: Formula/step.rb
Comment (dopey) [reply Formula/step.rb]: Yep, using a resource made way more sense and also simplified our build pipeline.

File: Formula/step.rb
Comment (zbeekman) [inline Formula/step.rb]: This line [is failing on Jenkins](https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/37823/version=high_sierra/testReport/junit/brew-test-bot/high_sierra/test_step/) with the message:

```
==> /usr/local/Cellar/step/0.8.4/bin/step ca init --address 127.0.0.1:8081 --dns 127.0.0.1 --password-file /private/tmp/step-test-20190212-3676-p935hj/password.txt --provisioner-password-file /private/tmp/step-test-20190212-3676-p935hj/password.txt --name homebrew-smallstep-test --provisioner brew
error allocating terminal: open /dev/tty: device not configured
Error: step: failed
Failed executing:
/usr/local/Homebrew/Library/Homebrew/formula.rb:1921:in `block in system'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1858:in `open'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1858:in `system'
```

**Comment by zbeekman** [comment]
@dopey Very thorough tests! We'll definitely know if/when this breaks...

Speaking of which, any thoughts on the test failure on Jenkins?

Once the failure is resolved, the commits should be squashed (fixed up really) and then I think it should be ready to ship, modulo edits/suggestions from other maintainers.

File: Formula/step.rb
Comment (dopey) [reply Formula/step.rb]: Yep, we're in the process of fixing that right now. Looks like we were opening up prompt terminals even when the prompt was satisfied via the cli. I should be pushing an update in 15-30 mins.

**Comment by dopey** [comment]
Appreciate the feedback and the help @zbeekman. I think we know what caused the jenkins failure and it actually led to us fixing a bug in our code. I'm going to build a release candidate of cli and push a formula with the rc url just so I can verify that the jenkins tests pass before cutting a standard release.

**Comment by dopey** [comment]
Pushed a new version and fixed up the commit log. Looking forward to hearing back from other maintainers.

**Comment by zbeekman** [comment]
I make no guarantees they’ll respond in a timely fashion, but after I look
this over again, it’ll get merged if no one spots any issues within a
reasonable time frame.

**Review by zbeekman — APPROVED** [review]
You know you've met a true professional when the tests far outnumber the lines of actual source code! Thanks for your contribution, @dopey!

LGTM! I vote we :ship: it!

File: Formula/step.rb
Comment (zbeekman) [reply Formula/step.rb]: Holy tests, Batman! 🥇

**Comment by zbeekman** [comment]
HMMM, I need to figure out what's up with the bottle/bintray.

In the meantime, @dopey, would you mind amending the commit and force pushing it?

The commit text should read:

```
step 0.8.5 (new formula)

# If you want to you can add stuff here, a description of step and step-ca would be good
# i.e., you can keep the first two bullets of your original commit
```

I'm hoping that after you amend the commit message and force push whatever bottling issue I just experienced will be resolved.

**Comment by zbeekman** [comment]
Thanks for your contribution to Homebrew, @dopey!

**Comment by dopey** [comment]
"Noted" on the commit message for future contributions. Thanks for the guidance!

**Comment by dopey** [comment]
Hey @zbeekman, while testing locally I found that the executables (`step` and `step-ca`) were missing versions. So `step version` would return 

```
$ step version
Smallstep CLI/ (darwin/amd64)
Release Date: 2019-02-13 22:58 UTC
```

rather than ...

```
$ step version
Smallstep CLI/0.8.5 (darwin/amd64)
Release Date: 2019-02-13 22:58 UTC
```

And same for `step-ca`. The fix is pretty simple, just involves setting a `VERSION` env var before running make for each command. I've made that patch and added a few homebrew tests to make sure we don't regress, but I'm no longer able to push to this PR b/c the PR is closed.

Would you recommend we re-open the PR, or submit a new PR fixing the issue once this PR has been merged?

**Comment by zbeekman** [comment]
> Would you recommend we re-open the PR, or submit a new PR fixing the issue once this PR has been merged?

Hi @dopey, please open a new PR with your proposed changes.

Due to our workflow, the merging of the PR actually happens as a direct commit to master using the magic of `brew pull`. It updates/inserts the bottle-do block, and ensures that the bottles are staged to bintray and available for download. When we do this, the commit that performs the "merge" uses the GH keyword to close the PR, indicating that it was effectively "merged". GH automatically links that commit in the PR discussion stream, you can see it below [this comment](https://github.com/Homebrew/homebrew-core/pull/36007#issuecomment-463161427). The merge happened in https://github.com/Homebrew/homebrew-core/commit/c936b8375577e0da2870540f64474e76f9931cd5.

**Pull request closed without merging**
